enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -2795,46 +2841,34 @@ func (e *partitionEvictor) sample(ctx context.Context, k int) ([]*approxlru.Samp
 		LowerBound: start,
 		UpperBound: end,
 	})
-	iter.SeekGE(start)
 	defer iter.Close()
 
-	if !*acEvictionEnabled {
-		e.sampleGroup()
-	}
-
+	// Files are kept in random order (because they are keyed by digest), so
+	// instead of doing a new seek for every random sample we will seek once
+	// and just read forward, yielding digests until we've found enough. To
+	// ensure we sample all kinds of items, e.lastSampledType  will be
+	// toggled on every call to sample.
 	samples := make([]*approxlru.Sample[*evictionKey], 0, k)
+	var key filestore.PebbleKey
 
-	// generate k random digests and for each:
-	//   - seek to the next valid key, and return that file record
+	// Generate k random samples. Attempt this up to k*2 times, to account
+	// for the fact that some files sampled may be younger than
+	// minEvictionAge. We try hard!
 	for i := 0; i < k*2; i++ {
-		groupID := ""
-		cacheType := rspb.CacheType_CAS
-		if i%2 == 1 && *acEvictionEnabled {
-			gid, err := e.randomGroupForEvictionSampling()
-			if err == nil {
-				cacheType = rspb.CacheType_AC
-				groupID = gid
-			} else if !status.IsNotFoundError(err) {
-				log.Warningf("no groups to sample for %q: %s", e.part.ID, err)
+		if !iter.Valid() {
+			// This should only happen once per call to sample(), or
+			// occasionally more if we've exhausted the iter.
+			randomKey, err := e.randomKeyForEvictionSampling()
+			if err != nil {
+				return nil, err
 			}
+			iter.SeekGE(randomKey)
```

Comment:
Not sure if this was intentional but you are skipping the first key since you call SeekGE and then immediately call iter.Next(). Shouldn't really matter in practice since we're picking random keys anyway, but might be worth documenting that this is intentional.

Reply:
PTAL

Reply:
looks good